# Code for "Characterization of cortical neurodevelopment in vitro using gene expression and morphology profiles from single cells"

Currently available at [BioArxiv](https://doi.org/10.1101/2023.12.03.569583]).

## List of R/python dependencies:

### R
- oob (https://github.com/DimitriMeistermann/oob)
- GSDS (https://github.com/DimitriMeistermann/GSDS)
- DBI
- plotly
- igraph
- Seurat
- wesanderson
- ggpubr
- gridExtra
- grid
- gridtext
- scater
- scales
- readxl
- ggplot2
- dplyr
- harmony
- GenomicRanges
- patchwork
- FNN
- glmnet
- ggbeeswarm
- ggrepel
- stringr
- ComplexHeatmap

### python
- igraph
- pip
- pandas
- scikit-learn
- umap-learn
- jupyter
- matplotlib
- seaborn
- openpyxl
- numpy
- leidenalg
- trimap

## Rerun R/python code

### Cell Painting

1. Execute `git clone https://github.com/Kilpinen-group/cortical_diff_code.git`
2. Download the zenodo [archive](https://zenodo.org/records/13913047), extract and rename the folder *zenodoArchive*
3. Copy the zenodo folder in the repository directory (cortical_diff_code).
4. Install the dependencies
5. Run the code, do not forget to change the `setwd` lines.

### scRNA-Seq
TBA

### scRNA-Seq X Cell Painting
TBA